runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -3304,3 +3304,82 @@ J9::ValuePropagation::innerConstrainAcall(TR::Node *node)
 
    return node;
    }
+
+TR_OpaqueClassBlock *
+J9::ValuePropagation::findLikelySubtype(TR_OpaqueClassBlock *klass)
+   {
+   if (!klass || TR::VPConstraint::isSpecialClass((uintptr_t)klass))
+      return NULL;
+
+   int32_t numDims = 0;
+   klass = ((TR_J9VM *)fe())->getBaseComponentClass(klass, numDims);
+
+   // If the returned base component class is still an array, it's a primitive array
+   if (TR::Compiler->cls.isClassArray(comp(), klass))
+      return NULL; // primitive array, should be fixed-type anyway
+
+   if (TR::Compiler->cls.isInterfaceClass(comp(), klass)
+       || TR::Compiler->cls.isAbstractClass(comp(), klass))
+      {
+      TR::ClassTableCriticalSection lock(fe());
+      auto *chTable = comp()->getPersistentInfo()->getPersistentCHTable();
+      klass = chTable->findSingleConcreteSubClass(klass, comp());
+      }
+   // If it's a concrete class and it has been extended, there is no particularly likely type
+   else if (fe()->classHasBeenExtended(klass)
+            || TR::Compiler->vm.isVMInStartupPhase(comp()))
+      {
+      return NULL;
+      }
+
+   while (klass != NULL && numDims > 0)
+      {
+      klass = fe()->getArrayClassFromComponentClass(klass);
+      numDims--;
+      }
+
+   klass = (numDims == 0) ? klass : NULL;
```

Comment:
There's no need to check `numDims` here. On exit from the loop, we know that `klass == NULL || numDims == 0`, so if `numDims != 0`, then `klass == NULL`. Not that it really hurts to check `numDims` either, so if you really want to keep it, that's fine too

Similarly in the signature-based overload

Reply:
I see now after the while loop, the check is redundant. I'll remove it in the next commit

Reply:
Addressed in 4f8de01